### PR TITLE
stream: shift websocket reader and dialer code from wrapper to stream package

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

~~Opened as draft for initial inspection and critique.~~ This attempts to reduce code duplication for websocket wrappers and gets it another step forward for multi-connection handling. 

* Define websocket processor handlers per individual connection - This will eventually be bespoke connections only handling specific assets and or constrained subs. 
* Define bootstrap function for bespoke connection startup code per individual connection
* Shift connection dialing and session handling away from wrappers 
* Shift websocket reader routines generation to stream package 
* Remove proxy field websocket connection as its handled 1 proxy per session anyway 
* Adds basic (non-tls) mock server and mock proxy for websocket testing
* Remove superfluous running URL fields as connections can differ
* Adds new currency error `ErrCurrencyPairRequired`
* Split `SetWebsocketURL` into two functions `SetWebsocketURL` && `SetWebsocketAuthURL`
* Split `GetWebsocketURL` into two functions `GetWebsocketURL` && `GetWebsocketAuthURL`
* Removes unused websocket field `DefaultURL`
* Removes unexported methods `setState`, `setEnabled`, `setDataMonitorRunning`, `dataMonitorRunning`, `checkAndSetMonitorRunning`, `setTrafficMonitorRunning`, `setConnectionMonitorRunning` and use the field methods directly. 
* Defered reader.Close in `parseBinaryResponse()` so as to make sure if `io.ReadAll` fails it will still close.

Websocket_Connection

* Remove `setConnectedStatus` and `IsConnected` methods in favour of field methods. 

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/stream/websocket.go [hunk 1]
```diff
@@ -285,20 +277,48 @@ func (w *Websocket) Connect() error {
...
-		return fmt.Errorf("%v Error connecting %w", w.exchangeName, err)
+	if w.connector != nil {
+		err := w.connector()
+		if err != nil {
+			w.setState(disconnected)
+			return fmt.Errorf("%v Error connecting %w", w.exchangeName, err)
+		}
+	}
+
+	if w.Conn != nil {
```
Comment (gloriousCode) [inline exchanges/stream/websocket.go:288]: This setup doesn't work. Non-new websockets will panic here as w.Conn is defined, but the functions below are not

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x104adf5a8]

goroutine 120 [running]:
github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*Websocket).listen(0x140001ae3c0, {0x1058830d8, 0x140008e0100}, 0x0)
        /Users/scottg/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/stream/websocket.go:992 +0x78
created by github.com/thrasher-corp/gocryptotrader/exchanges/stream.(*Websocket).InitialiseConnection in goroutine 146
        /Users/scottg/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/stream/websocket.go:1022 +0x1a8
```

This is for Kraken, it shouldn'tbe hitting this new code for Kraken, right? No errors are thrown from nil functions in `InistialiseConnection`

File: exchanges/stream/websocket_types.go [hunk 2]
```diff
@@ -93,9 +92,14 @@ type Websocket struct {
...
 
 	// Standard stream connection
-	Conn Connection
+	Conn            Connection
+	UnAuthHandler   func([]byte) error
+	UnAuthBootstrap func(Connection) error
+
 	// Authenticated stream connection
-	AuthConn Connection
+	AuthConn      Connection
```
Comment (gloriousCode) [inline exchanges/stream/websocket_types.go:100]: Down the line, I would love for AuthConn to be removed. An auth connection is just another connection. At present, any connection issues for unauth/auth affect the other and it doesn't make much sense for the direction we're heading with multiple websocket connection types

**Review by gloriousCode — CHANGES_REQUESTED** [review]
I like the concepts. Will need some more work, but defining functions at the Connection Setup level is nice. The removal of needing duplicate ReadMessage functions is nice

File: exchanges/okx/okx_wrapper.go [hunk 3]
```diff
@@ -225,36 +223,41 @@ func (ok *Okx) Setup(exch *config.Exchange) error {
...
-	if err := ok.Websocket.SetupNewConnection(stream.ConnectionSetup{
+	if err := ok.Websocket.SetupNewConnection(&stream.ConnectionSetup{
 		URL:                  okxAPIWebsocketPublicURL,
 		ResponseCheckTimeout: exch.WebsocketResponseCheckTimeout,
 		ResponseMaxLimit:     okxWebsocketResponseMaxLimit,
 		RateLimit:            500,
+		Handler:              ok.WsHandleData,
+		Bootstrap:            ok.WsUnAuthBootstrap,
+		ReadBufferSize:       8192,
+		WriteBufferSize:      8192,
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/okx/okx_wrapper.go:246]: Yeah that is something that I want to think about implementing after shifting subscription handling over to individual connections. So we can hit a max ceiling and start generating new connections to handle everything.

File: exchanges/stream/websocket.go [hunk 4]
```diff
@@ -1009,3 +980,45 @@ func (w *Websocket) checkSubscriptions(subs []subscription.Subscription) error {
...
+		if err := handler(resp.Raw); err != nil {
+			w.DataHandler <- err
+		}
+	}
+}
+
+// InitialiseConnection sets up a websocket connection
+func (w *Websocket) InitialiseConnection(conn Connection, bootstrap func(Connection) error, handler func([]byte) error) error {
+	dialer := *websocket.DefaultDialer
+	if w.proxyAddr != nil {
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/stream/websocket.go:1001]: I un exported it and only call it in `Websocket Connect` method which is locked down.

File: exchanges/stream/websocket.go [hunk 5]
```diff
@@ -677,76 +680,70 @@ func (w *Websocket) CanUseAuthenticatedWebsocketForWrapper() bool {
...
+		log.Debugf(log.WebsocketMgr, "%s websocket: flushing websocket connection to %s\n", w.exchangeName, url)
 		return w.Shutdown()
 	}
 	return nil
 }
 
 // GetWebsocketURL returns the running websocket URL
-func (w *Websocket) GetWebsocketURL() string {
-	return w.runningURL
+func (w *Websocket) GetWebsocketURL(auth bool) string {
```
Comment (gbjk) [inline exchanges/stream/websocket.go:720]: Two issues with this:
* I don't think we should go further down the "websocket is a websocket but maybe actually 2 and one of them is auth" route.
* I don't think there's any cleaner API here in passing a boolean than having 2 methods

So in both cases I vote for GetWebsocketAuthURL() which has convenience handling for AuthConn is nil.

File: exchanges/stream/websocket.go [hunk 6]
```diff
@@ -677,76 +680,70 @@ func (w *Websocket) CanUseAuthenticatedWebsocketForWrapper() bool {
 
 // SetWebsocketURL sets websocket URL and can refresh underlying connections
 func (w *Websocket) SetWebsocketURL(url string, auth, reconnect bool) error {
```
Comment (gbjk) [inline exchanges/stream/websocket.go:682]: Back to front-review-comments 🤦 
See my comment below on GetWebsocketURL.
I actually vote for 2 methods on this too, even though it changes the API.

Trying to think about how this pattern would look like for a router, but I'm about a week away from putting that up on the drawing board. Still mulling it over.

File: exchanges/stream/websocket.go [hunk 7]
```diff
@@ -604,64 +636,35 @@ func (w *Websocket) trafficMonitor() {
...
+func (w *Websocket) setDataMonitorRunning(b bool)       { w.dataMonitorRunning.Store(b) }
+func (w *Websocket) checkAndSetMonitorRunning() (alreadyRunning bool) {
+	return !w.connectionMonitorRunning.CompareAndSwap(false, true)
 }
 
 // IsInitialised returns whether the websocket has been Setup() already
-func (w *Websocket) IsInitialised() bool {
-	return w.state.Load() != uninitialised
-}
+func (w *Websocket) IsInitialised() bool { return w.state.Load() != uninitialised }
```
Comment (gbjk) [inline exchanges/stream/websocket.go:649]: Just a note on all of these one-linifications:
I've often considered them non-idiomatic, since I've never seen them in the core lib or written by any of the trend setters.
I also find them difficult to parse by scanning, because the content indentation varies if it has a comment, which public funcs need to have.

So I guess I'm saying there should be a space after 644 to make `checkAndSetMonitorRunning` body parsable, and that 649-677 should be multi-liners.

This is subjective. But if you were on the fence or debating it, that's my taste on it.

File: exchanges/bitfinex/bitfinex.go [hunk 8]
```diff
@@ -1688,7 +1688,7 @@ func (b *Bitfinex) CancelMultipleOrdersV2(ctx context.Context, orderID, clientOr
 					cancelledOrder.AuxLimitPrice = f
 				}
 			}
-			cancelledOrders[y] = cancelledOrder
+			cancelledOrders = append(cancelledOrders, cancelledOrder)
```
Comment (gbjk) [inline exchanges/bitfinex/bitfinex.go:1691]: Does this fix something?
In any case should be in it's own commit, and if it's a fix PR'd separately.
If it's fixing something, should have a test to prove it was broken before too 😄

File: exchanges/exchange_test.go [hunk 9]
```diff
@@ -208,7 +208,7 @@ func TestSetClientProxyAddress(t *testing.T) {
 		t.Error("SetClientProxyAddress parsed invalid URL")
 	}
 
-	if newBase.Websocket.GetProxyAddress() != "" {
+	if newBase.Websocket.GetProxyAddress().String() != "" {
```
Comment (gbjk) [inline exchanges/exchange_test.go:211]: Think that this would work with assert.NotEmpty
And we should be explicit about why we're expecting this, e.g. "SetClientProxyAddress should not set ProxyAddress on error"

File: exchanges/exchange_test.go [hunk 10]
```diff
@@ -223,7 +223,7 @@ func TestSetClientProxyAddress(t *testing.T) {
 		t.Error("trying to set the same proxy addr should thrown an err for ws")
 	}
 
-	if newBase.Websocket.GetProxyAddress() != "http://www.valid.com" {
+	if newBase.Websocket.GetProxyAddress().String() != "http://www.valid.com" {
```
Comment (gbjk) [inline exchanges/exchange_test.go:226]: Ditto

File: exchanges/okx/okx_test.go [hunk 11]
```diff
@@ -988,7 +988,7 @@ func TestCancelAllQuotes(t *testing.T) {
 	switch {
 	case err != nil:
 		t.Error("Okx CancelAllQuotes() error", err)
-	case err == nil && time.IsZero():
+	case time.IsZero():
```
Comment (gbjk) [inline exchanges/okx/okx_test.go:991]: assert.Empty()

File: exchanges/okx/okx_test.go [hunk 12]
```diff
@@ -2264,7 +2264,7 @@ func TestGetOrderHistory(t *testing.T) {
 	_, err := ok.GetOrderHistory(contextGenerate(), &getOrdersRequest)
 	if err == nil {
 		t.Errorf("Okx GetOrderHistory() Expected: %v. received nil", err)
-	} else if err != nil && !errors.Is(err, errMissingAtLeast1CurrencyPair) {
+	} else if !errors.Is(err, errMissingAtLeast1CurrencyPair) {
 		t.Errorf("Okx GetOrderHistory() Expected: %v, but found %v", errMissingAtLeast1CurrencyPair, err)
 	}
```
Comment (gbjk) [inline exchanges/okx/okx_test.go:2269]: This makes the 2265 test redundant, since nil is not errMissingAtLeast1CurrencyPair.

Suggest the whole thing becomes:
```suggestion
	assert.ErrorIs(t, err, errMissingAtLeast1CurrencyPair, "GetOrderHistory should error when not passed any pair")
```

DriveBy: That's a horrible error message name. Sounds like errPairRequired to me, and we probably should have a common.Err for that, I'd imagine. Or currency.Err

File: exchanges/okx/okx_websocket.go [hunk 13]
```diff
@@ -213,86 +213,41 @@ const (
...
-		var authDialer websocket.Dialer
-		authDialer.ReadBufferSize = 8192
-		authDialer.WriteBufferSize = 8192
-		err = ok.WsAuth(context.TODO(), &authDialer)
-		if err != nil {
-			log.Errorf(log.ExchangeSys, "Error connecting auth socket: %s\n", err.Error())
-			ok.Websocket.SetCanUseAuthenticatedEndpoints(false)
-		}
-	}
+// WsUnAuthBootstrap bootstraps the unauthenticated connection
```
Comment (gbjk) [inline exchanges/okx/okx_websocket.go:218]: Think this is just `WsBootstrap`
I'm Starting to see how this would work now.

* One WsBootstrap
* Passed a stream.Connection
* stream.Connection has a type flag
* Exchange can delegate internally based on that

File: exchanges/stream/websocket.go [hunk 14]
```diff
@@ -285,15 +297,39 @@ func (w *Websocket) Connect() error {
...
+	if w.connector != nil {
+		err := w.connector()
+		if err != nil {
+			w.setState(disconnected)
+			return fmt.Errorf("%v Error connecting %w", w.exchangeName, err)
+		}
+	}
+
+	if w.connector == nil && w.Conn != nil {
+		err := w.initialiseConnection(w.Conn, w.UnAuthBootstrap, w.UnAuthHandler)
```
Comment (gbjk) [inline exchanges/stream/websocket.go:309]: Suggest `initConnection`

File: exchanges/stream/websocket.go [hunk 15]
```diff
@@ -1009,3 +996,46 @@ func (w *Websocket) checkSubscriptions(subs []subscription.Subscription) error {
...
+// listen listens to the websocket connection and handles incoming data
+func (w *Websocket) listen(conn Connection, handler func(incoming []byte) error) {
+	defer w.Wg.Done()
+	for {
+		resp := conn.ReadMessage()
+		if resp.Raw == nil {
+			return
+		}
+		if err := handler(resp.Raw); err != nil {
+			w.DataHandler <- fmt.Errorf("websocket streaming connection %s %s: %w", w.exchangeName, conn.GetURL(), err)
```
Comment (gbjk) [inline exchanges/stream/websocket.go:1009]: Suggest that along with adding conn Type flag, that's what we'd use here, maybe stringified, but maybe as a structured error, too.
I think the first part of this isn't clear.
```
errWsHandler = errors.New("error in WS data handler")
```
```
"%s: %w %s: %w", eName, errWsHandler, conn.Type, err
```
`bybit: error in WS data handler public: semi-pornographic ascii art received instead of order confirmation`
or maybe as far as
```
package websocket
type HandlerError struct {
 Exchange exchange.IBotExchange (pointer obvs)
 Connection 
 Message ( the thing that caused this )
 Error
}

File: exchanges/stream/websocket.go [hunk 16]
```diff
@@ -1009,3 +996,46 @@ func (w *Websocket) checkSubscriptions(subs []subscription.Subscription) error {
...
+	if w.proxyAddr != nil {
+		// Note: This is a global setting and will affect all websocket
+		// connections. If you need to use a proxy for a single connection you
+		// will need to create a new dialer and set the proxy on that dialer.
+		// This reduces the shared state.
+		dialer.Proxy = http.ProxyURL(w.proxyAddr)
+	}
+
+	err := conn.Dial(&dialer, nil)
+	if err != nil {
```
Comment (gbjk) [inline exchanges/stream/websocket.go:1027]:
```suggestion
	if err := conn.Dial(&dialer, nil); err != nil {
```

File: exchanges/stream/websocket.go [hunk 17]
```diff
@@ -1009,3 +996,46 @@ func (w *Websocket) checkSubscriptions(subs []subscription.Subscription) error {
...
+		return fmt.Errorf("%v connecting websocket: %w", w.exchangeName, err)
+	}
+
+	if bootstrap != nil {
+		err = bootstrap(conn)
+		if err != nil {
```
Comment (gbjk) [inline exchanges/stream/websocket.go:1033]:
```suggestion
		if err = bootstrap(conn); err != nil {
```

File: exchanges/stream/websocket_test.go [hunk 18]
```diff
@@ -106,10 +101,87 @@ func (d *dodgyConnection) Connect() error {
...
+				messageType, message, err := actualServerConn.ReadMessage()
+				if err != nil {
+					return
+				}
+				if err := conn.WriteMessage(messageType, message); err != nil {
+					return
+				}
+			}
+		}()
+	}))
```
Comment (gbjk) [inline exchanges/stream/websocket_test.go:176]: Should be a func. Or 2.

File: exchanges/stream/websocket_test.go [hunk 19]
```diff
@@ -687,10 +777,10 @@ func TestSendMessage(t *testing.T) {
...
 		{Error: nil,
 			WC: WebsocketConnection{
-				ExchangeName:     "test3",
-				Verbose:          true,
-				URL:              websocketTestURL,
-				ProxyURL:         proxyURL,
+				ExchangeName: "test3",
+				Verbose:      true,
+				URL:          websocketTestURL,
+				// ProxyURL:         proxyURL,
```
Comment (gbjk) [inline exchanges/stream/websocket_test.go:783]: ☝️ Feels like *maybe* these were leaked, so just highlighting.

File: exchanges/stream/websocket_test.go [hunk 20]
```diff
@@ -699,7 +789,7 @@ func TestSendMessage(t *testing.T) {
 		testData := &testCases[i]
 		t.Run(testData.WC.ExchangeName, func(t *testing.T) {
 			t.Parallel()
-			if testData.WC.ProxyURL != "" && !useProxyTests {
+			if /*testData.WC.ProxyURL != "" &&*/ !useProxyTests {
```
Comment (gbjk) [inline exchanges/stream/websocket_test.go:792]: ☝️ Feels like *maybe* these were leaked, so just highlighting.

File: exchanges/stream/websocket_types.go [hunk 21]
```diff
@@ -107,8 +113,8 @@ type Websocket struct {
 
 // WebsocketSetup defines variables for setting up a websocket connection
 type WebsocketSetup struct {
-	ExchangeConfig        *config.Exchange
-	DefaultURL            string
+	ExchangeConfig *config.Exchange
+	// DefaultURL            string
```
Comment (gbjk) [inline exchanges/stream/websocket_types.go:117]: ☝️ Feels like *maybe* these were leaked, so just highlighting.

**Review by gbjk — COMMENTED** [review]
Great work 🎉 
Subjective comments mostly

I think it moves us in the right direction, definitely.
I'm mindful of what our overall websockets structure will look like, and the current mixing of naming `stream`, `websocket` and `connection`, but I think this improves, generally.

File: exchanges/stream/websocket.go (hunk 7 above)
Comment (gloriousCode) [reply exchanges/stream/websocket.go:649]: I didn't enjoy them either, if that helps sway you Shazbert 😄

Comment (shazbert) [reply exchanges/stream/websocket.go:649]: [swayed](https://youtu.be/a90tZJHBklk)

File: exchanges/stream/websocket_connection.go [hunk 22]
```diff
@@ -294,21 +241,21 @@ func (w *WebsocketConnection) Shutdown() error {
 	if w == nil || w.Connection == nil {
 		return nil
 	}
-	w.setConnectedStatus(false)
-	return w.Connection.UnderlyingConn().Close()
+	w.connected.Store(false)
+	return w.Connection.NetConn().Close()
```
Comment (shazbert) [inline exchanges/stream/websocket_connection.go:245]: UnderlyingConn was deprecated so its now `NetConn()`

**Comment by shazbert** [comment]
Closing in favour of #1580.
